src/prog_models/surrogates/dmd.py

```diff
@@ -94,31 +99,43 @@ def __init__(self, m, load_functions, **kwargs):
             time_data_interp = np.arange(times[0], times[-1], config['save_freq'])
 
             states_data_interp = {}
-            inputs_data_interp = {}
+            outputs_data_interp = {}
+            es_data_interp = {}
 
             for state_name in m.states:
                 states_data_temp = [states[iter_data1][state_name] for iter_data1 in range(len(states))]
                 states_data_interp[state_name] = interp1d(times,states_data_temp)(time_data_interp)
-            for input_name in m.inputs:
-                inputs_data_temp = [inputs[iter_data4][input_name] for iter_data4 in range(len(inputs))]
-                inputs_data_interp[input_name] = interp1d(times,inputs_data_temp)(time_data_interp)
+                if config['training_noise'] != 0:
+                    states_data_interp[state_name] += np.random.randn(len(states_data_interp[state_name]))*config['training_noise']
+            for output_name in m.outputs:
+                outputs_data_temp = [outputs[iter_data3][output_name] for iter_data3 in range(len(outputs))]
+                outputs_data_interp[output_name] = interp1d(times,outputs_data_temp)(time_data_interp)
+                if config['training_noise'] != 0:
+                    outputs_data_interp[output_name] += np.random.randn(len(outputs_data_interp[output_name]))*config['training_noise']
+            for es_name in m.events:
+                es_data_temp = [event_states[iter_data2][es_name] for iter_data2 in range(len(event_states))]
+                es_data_interp[es_name] = interp1d(times,es_data_temp)(time_data_interp)
+                if config['training_noise'] != 0:
+                    es_data_interp[es_name] += np.random.randn(len(es_data_interp[es_name]))*config['training_noise']
```

Comment:
avoid if? Just use:
states_data_interp[state_name] += np.random.randn(len(states_data_interp[state_name]))*config['training_noise']

Reply:
That will remove a line of code, which is nice. But the way Katy implemented it now will be much more efficient when time is 0. 

I timed it to illustrate. 
Case 1: training_noise is 0
   with if statement: 0.029 micro seconds per event
   without if statement: 3.8 micro seconds per event
Case 2: training_noise is not 0
   with if statement: 3.96 micro seconds per event
    without if statement: 3.8 micro seconds per event
    
Since it's not in the inner loop of simulation - I would accept it either way. It will save 10s of micro seconds in creating the surrogate if there's no noise, while only adding ~1s of microseconds if there is noise. I don't think it hurts to have this micro optimization.

Reply:
Here's my code if anyone wants to recreate it

`
>>> import time
>>> f1 = """\
... a += np.random.randn(len(a))*b
... """
>>> f2 = """\
... if b != 0:
...    a += np.random.randn(len(a))*b
... """
>>> timeit.timeit(f1, setup="import numpy as np; b=0; a = np.array([[1.0, 2.2],[2.1, 3.1]])")
>>> timeit.timeit(f2, setup="import numpy as np; b=0; a = np.array([[1.0, 2.2],[2.1, 3.1]])")
>>> timeit.timeit(f1, setup="import numpy as np; b=1; a = np.array([[1.0, 2.2],[2.1, 3.1]])")
>>> timeit.timeit(f2, setup="import numpy as np; b=1; a = np.array([[1.0, 2.2],[2.1, 3.1]])")
`

Reply:
Really interesting results! Thanks for running that. I agree that both options have pros and cons based on speed and likely frequency. For simplicity sake, I think I'll leave it as-is for now, and we'll gain the small micro-optimization. Thanks to both of you for your perspective on this.